File: IPv4 Validation.php
# IPv4 Validation PHP
# Author: Sandro Zappulla
# Github: @SandroZappulla
# version: v1.0

// bedingung: Wenn submittet wurde
if ($_SERVER["REQUEST_METHOD"] == "POST"){
    // VARIABLES
    $ipv4 = htmlspecialchars($_POST['ipv4']);

    echo "Objekt_name: ", $object_name, "<br>";
    echo "IPv4: ", $ipv4, "<br>";
    echo "IPv6: ", $ipv6;
    echo "<br><br><br><br><br>";


    // VALIDATION

    // *** IPV4-ADDRESS ***
    // error_msg
    $error_empty = "Eingabe darf nicht leer sein!<br>";
    $error_ipv4_pattern = "Das Pattern der IPv4-Adresse stimmt nicht ein!<br>";
    $error_length = "Die Länge muss mind. 7 Zeichen und max 15 Zeichen Lang sein!<br>";
    $error_first_char_is_digit = "Erstes Zeichen muss eine Zahl enthalten!<br>";
    $error_digits_or_dots = "Nur Zahlen oder Punkte Verwenden!<br>";
    $error_amount_dots = "Es müssen exakt 3 Punkte enthalten sein.<br>";
    $error_unvalid_digit = "Eine Verwendetes Oktett besitzt nicht eine Zahlen range von (0-255)! <br>";

    // Counter
    $char_counter = 0;

    // Überprüfe: Ob Leerer Eintrag (Auch für Objektname und IPv6 Adresse)
    if(empty($object_name) || (empty($ipv4)) || (empty($ipv6))) {
        echo $error_empty;
    }



    // Überprüfe: Länge
    $ipv4_length = strlen($ipv4);
    if ($ipv4_length < 7 || $ipv4_length > 15){
        echo $error_length;
    }



    // Überprüfe: Anzahl der Punkte
    $dot_count = substr_count($ipv4,".");

    if ($dot_count != 3){
        echo $error_amount_dots;
    }



    // Überprüfe: Zahl oder Punkt vorhanden
    do{
        if (is_numeric($ipv4[$char_counter]) || $ipv4[$char_counter] == "."){
        }
        else {
            echo $error_digits_or_dots;
        }

        $char_counter+=1;  
    } while($char_counter < $ipv4_length);



    // RegEx Pattern Überprüfen
    // Pattern 
    $ipv4_pattern = '/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/';
    
    if (preg_match($ipv4_pattern, $ipv4) == 0){ // Rückgabewert bei preg_match() ist 0 wenn Falsch und 1 wenn richtig
        echo $error_ipv4_pattern;
    }



    // Überprüfe ob erstes Zeichen eine Zahl ist
    $first = $ipv4[0];
    if (!is_numeric($first)){
        echo $error_first_char_is_digit;
    }

    // Überprüfung: Validierung einer einzelnen Zahl (Zwischen 0-255)
    $oktett = explode(".", $ipv4);

    $counter_single_digit = 0;

    do {
        if ($oktett[$counter_single_digit] < 0 || $oktett[$counter_single_digit] > 255){
            $unvalid_digit +=1;
            echo $error_unvalid_digit;
            break;
        }
        $counter_single_digit +=1;
    } while ($counter_single_digit < $dot_count+1);
}
// Wenn nicht submittet wurde direkt zurücksenden zur index.php #SECURITY
else {
    header("Location: index.php");
}

File: IPv4 Validation.py
# IPv4 Validation
# Author: Sandro Zappulla
# Github: @SandroZappulla
# version: v1.0

# IMPORT
# RegEx
import re

# Variables # TestOnly
#my_ip = "192.168.178.177"
#test = ""
#short_ip = "1.1.1.1"
#not_valid_ip = "256.-1.1000.532"

# Eintrag # TestOnly
#IPv4_entry = my_ip

# checks
check_empty_input = False
check_length = False
check_valid_chars = True
check_first_char_is_digit = False
check_amount_dots = False
check_regex_pattern = True
check_number_range = False



# MAIN
print("IPv4 Validierung by Sandro Zappulla")
IPv4_entry = input("Bitte Geben Sie eine gültige IPv4-Adresse ein: ") # Eingabe

# Überprüfe: Nicht leer
if IPv4_entry == '':
    check_empty_input = True



# Überprüfe: Länge (Mind. 7 Max. 15 Zeichen)
length = len(IPv4_entry)
if length >= 7 and length <= 15:
    check_length = True 



# Überprüfe: Ob NUR Ziffern und Punkte enthalten sind
char_counter = 0
valid_chars = 0
unvalid_chars = 0

while char_counter < length:
    # Überprüfung ob beim index von der Eingabe eine Zahl vorhanden ist oder ein "."
    if IPv4_entry[char_counter].isdigit() or IPv4_entry[char_counter] == '.':
        valid_chars+=1
    else:
        unvalid_chars+=1
    char_counter+=1

if unvalid_chars != 0:
    check_valid_chars = False



# Überprüfe: Erstes Zeichen Zahl
first = IPv4_entry[0]

if first.isdigit():
    check_first_char_is_digit = True



# Überprüfe: Anzahl der Punkte
dot_count = IPv4_entry.count('.')

if dot_count == 3:
    check_amount_dots = True
elif dot_count != 3:
    check_amount_dots = False



# Überprüfe: Keine Punkte hintereinander
# RegEx Überprüfung: Vorgabe(1-3 Zahlen, dann darauffolgender ".") (^ = Beginn, $ = Ende)
pattern = re.compile("^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
if pattern.match(IPv4_entry) == None:
    check_regex_pattern = False


# Oktett Bildung: Unterteilung von Punkt in einer Liste
digit_split = IPv4_entry.split(".")



# Überprüfung: Validierung einer einzelnen Zahl (Zwischen 0-255)
correct_number_counter = 0

# Bevor jede Zahl einzelnd Validiert wird erwarten wir das erst die Überprüfung "Keine Punkte Hintereinander (RexEx)" richtig ist
if check_regex_pattern == True:
    for digit in digit_split:
        int_digit = int(digit)
        #print(int_digit)

        if int_digit >= 0 and int_digit <= 255:
            correct_number_counter+=1
    
        if correct_number_counter == 4:
            check_number_range = True

# AUSGABE
print("------------------------------------- EINTRAG ----------------------------------------")
print("\n")
print(f"Eintrag: {IPv4_entry}")
print("\n")
print("------------------------------------- ÜBERPRÜFUNG ----------------------------------------")
print(f"check_empty_input: {check_empty_input}")
print(f"check_length: {check_length} : length: {length}")
print(f"check_valid_chars: {check_valid_chars} : valid_chars: {valid_chars} : unvalid_chars: {unvalid_chars}")
print(f"check_first_char_is_digit: {check_first_char_is_digit} : first: '{first}',")
print(f"check_amount_dots: {check_amount_dots} : dot_count: '{dot_count}',")
print(f"check_regex_pattern: {check_regex_pattern}")
print(f"check_number_range: {check_number_range} : correct_number_counter: {correct_number_counter}")
